votes/0150-pr-issue-association.md

```diff
@@ -0,0 +1,51 @@
+# A Vote for PR Issue Association
+
+* Discussion at https://github.com/tikv/community/pull/150
+* Tracking issue https://github.com/tikv/community/issues/149
+
+## Proposal
+
+### Motivation
+
+To measure software evolution, we focus on features implemented, bugs fixed, and general enhancement. Issues document these items, while a pull request focuses on the change set and whether we implement it correctly.
+
+Normally, an issue documents a work item, and a pull request resolves it (despite cherry-picks). However, the relation between pull requests and issues is complicated in TiDB projects.
+
+* 1 issue with 0 PR: question, bug vanished without a reason, etc. 
+* 1 issue with 1 PR: normal case. 
+* 1 issue with n PRs: resolve a bug multiple times ([1](https://github.com/pingcap/tidb/issues/24679)), or resolve a complex issue in steps([2](https://github.com/pingcap/tidb/issues/27652)).
+* n issues with 1 PR: trade off resolving things as a whole (due to code coupling, or we can resolve multiple bugs at the common parent level) ([3](https://github.com/pingcap/tidb/pull/27697)), duplicated issues. 
+* 0 issue with 1 PR: coding without a ticket ([4](https://github.com/pingcap/tidb/pull/23022)). Even if the description is detailed or even if a design document is present, pull requests focus on the change set. It is about the implementation, not the question; if it is, we always mix discussion because the change set has a primary position. 
+
+### Solution
+
+Previous discussion happened on the [developer discussion forum](https://internals.tidb.io/t/topic/409).
+
+I propose to check that a pull request must be associated with issue(s).
+
+In details:
+
+* A pull request MUST be associated with issue(s) in commit messages. That is, the change set by `git commit` with a message mention `ref #{issue-number}` or `close #{issue-number}` or semantic equivalent.
```

Comment:
OK. Then it is `ref #{issue-number}` or `close[sd]?|resolve[sd]?|fix(e[sd])? #{issue-numer}`. `ref` has no magic but for convention.

Reply:
> `ref` has no magic but for convention.

does that mean the tichi script mentioned below works for plain `#12345` without any keywords around?

Reply:
Resolved.

Reply:
@Mini256 should be more familiar with the technique part.

I tend to propose a happy path rather than tell every truth in details. Use `ref #xxx` when you don't want to close the issue and `close #xxx` when you do. Other ways may work if they actually associate a PR with an issue technically (for example, plain `#xxx`).

The statement here is "A pull request MUST be associated with issue(s) in commit messages". I can rewrite the following sentence in "You CAN achieve it by git commit` with a message mention..." if you ask for it.

Reply:
In terms of technology, I will provide a generic matching-blocker plugin through tichi, use the regex to check the content of the commit message, and the regex is configurable, for now, the robot only checks #xxx part.

@tisonkun, maybe you can add the example (pr: https://github.com/tikv/tikv/pull/10971, issue: https://github.com/tikv/tikv/issues/9012) provided by @BusyJay to the vote content.

It should be noted that in this scheme, **both commit and PR will be associated with the issue**. Especially considering that our repositories have multiple release branches, the issue page will generate more mentioned links.

Reply:
> both commit and PR will be associated with the issue

What do you mean exactly? If we don't check the PR body, it has nothing to do we mention "and PR" here, even if it _can_ be.